protocol/shannon/mode_centralized.go

```diff
@@ -39,17 +41,52 @@ func getCentralizedModeOwnedAppsAddr(ownedAppsPrivateKeysHex []string) ([]string
 	return ownedAppsAddr, nil
 }
 
-// getCentralizedGatewayModeAppFilter returns a permittedAppsFilter for the Centralized gateway mode.
-func getCentralizedGatewayModeAppFilter(gatewayAddr string, ownedAppsAddr map[string]struct{}) permittedAppFilter {
-	return func(app *apptypes.Application) error {
-		if _, found := ownedAppsAddr[app.Address]; !found {
-			return fmt.Errorf("Centralized GatewayMode: app with address %s is not owned by the gateway", app.Address)
+// appIsStakedForService returns true if the supplied application is staked for the supplied service ID.
+func appIsStakedForService(serviceID protocol.ServiceID, app *apptypes.Application) bool {
+	for _, svcCfg := range app.ServiceConfigs {
+		if protocol.ServiceID(svcCfg.ServiceId) == serviceID {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getCentralizedGatewayModeApps returns the set of permitted apps under the Centralized gateway mode.
+func (p *Protocol) getCentralizedGatewayModeApps(ctx context.Context, serviceID protocol.ServiceID) ([]*apptypes.Application, error) {
+	var permittedApps []*apptypes.Application
+
+	// Loop over the address of apps owned by the gateway in Centralized gateway mode.
+	for ownedAppAddr := range p.ownedAppsAddr {
+		onchainApp, err := p.FullNode.GetApp(ctx, ownedAppAddr)
+		if err != nil {
+			return nil, fmt.Errorf("Centralized GatewayMode: error getting onchain data for app %s owned by the gateway: %w", ownedAppAddr, err)
 		}
 
-		if !gatewayHasDelegationForApp(gatewayAddr, app) {
-			return fmt.Errorf("Centralized GatewayMode: app with address %s does not delegate to gateway address: %s", app.Address, gatewayAddr)
+		// Skip the app if it is not staked for the requested service.
+		if !appIsStakedForService(serviceID, onchainApp) {
+			// Logging is intentionally skipped here: this case is expected to occur frequently since
```

Comment:
Make it a debug log so we can enable it (or a warning). It's useful in debugging and understanding issues